[fix][broker] Fix ManagedCursor state management race conditions and lifecycle issues

lhotari opened the pull request:
Fixes #24566

### Motivation

The test `ReplicatorTest.testCloseReplicatorStartProducer` was consistently failing due to critical issues introduced in PR #24551. Investigation (shared in https://github.com/apache/pulsar/issues/24566#issuecomment-3119993989 by @3pacccccc ) revealed that the changes led to infinite recursive calls in `ManagedCursorImpl#cancelPendingReadRequest`, causing broker instability and test failures.

The root cause was improper state management for pending read operations, particularly when:
- Multiple threads attempt to cancel pending read requests concurrently
- Cursors are closed while read operations are pending
- OpReadEntry instances are recycled and reused incorrectly
- Cursor state transitions occur without proper synchronization
- Cursor deletion and closure operations race with active read operations

### Modifications

This PR addresses the state management issues through several key changes:

#### Core Synchronization Fixes

1. **Reverted problematic synchronization approach**: Removed the complex `pendingReadOpMutex` and `DelayCheckForNewEntriesTask` implementation added in #24551 that caused infinite recursion, returning to a more robust compare-and-swap based solution.

2. **Added operation ID tracking**: Introduced an `id` field to `OpReadEntry` instances to prevent recycled operations from being processed incorrectly. This ensures that only the correct operation instance is handled when checking for new entries.

3. **Improved atomic state transitions**: Enhanced the CAS (Compare-And-Swap) operations using `getAndUpdate()` with lambda functions that verify both the operation instance and its ID before making state changes:
   ```java
                OpReadEntry waitingReadOpItem = WAITING_READ_OP_UPDATER.getAndUpdate(this,
                        current -> {
                            if (current == op && current.id == opReadId) {
                                // update the value to null to cancel the waiting read op
                                return null;
                            } else {
                                // keep the current waiting read op value
                                return current;
                            }
                        });
                // If the waiting read op was the same as the one we are trying to cancel, it means that it was now
                // cleared from the waitingReadOp field and therefore "cancelled"
                if (waitingReadOpItem == op && waitingReadOpItem.id == opReadId) {
                    if (log.isDebugEnabled()) {
                        log.debug("[{}] [{}] Cancelled notification and scheduled read at {}", ledger.getName(),
                                name, op.readPosition);
                    }
                    PENDING_READ_OPS_UPDATER.incrementAndGet(this);
                    ledger.asyncReadEntries(op);
   ```

#### Enhanced State Management System

4. **Comprehensive state model enhancement**: Significantly expanded the cursor `State` enum with:
   - New states: `Deleting`, `Deleted`, `DeletingFailed` for better deletion lifecycle tracking
   - Added `closedState` boolean field and `isClosed()` method for more robust state checking
   - Added `isDeletingOrDeleted()` method for deletion state validation
   - Enhanced state constructor to properly categorize closed vs. open states

5. **Thread-safe state transition methods**: Implemented atomic state transition utilities:
   - `changeStateIfNotClosed()` - prevents state changes on closed cursors
   - `changeStateIfNotDeletingOrDeleted()` - prevents state changes on deleting/deleted cursors
   - Enhanced `trySetStateToClosing()` with atomic operations

#### Cursor Lifecycle Management

6. **Proper cursor closure handling**: Added comprehensive cursor closure logic:
   - `closeWaitingCursor()` - handles normal cursor closure scenarios
   - `cancelWaitingCursorsWhenDeactivated()` - handles cursor deactivation scenarios
   - `internalCloseWaitingCursor()` - unified internal implementation with exception supplier pattern
   - Added sentinel value (`WAITING_READ_OP_FOR_CLOSED_CURSOR`) to handle closed cursor scenarios

7. **Enhanced cursor deactivation**: Updated `setInactive()` to properly cancel waiting cursors when deactivated, preventing orphaned read operations.

8. **Robust deletion lifecycle**: Improved `asyncDeleteCursorLedger()` with:
   - Proper state transitions through deletion lifecycle
   - Enhanced retry logic with backoff for failed deletions
   - Better error handling and state recovery for deletion failures
   - Prevention of duplicate deletion attempts

#### Exception Handling Improvements

9. **New exception types**: Added `CursorDeactivatedWaitCallbackException` for better error propagation when cursors are deactivated without properly cancelling pending requests.

10. **Enhanced error propagation**: Improved error handling throughout the cursor lifecycle to ensure proper cleanup and user notification.

#### Managed Ledger Integration

11. **Enhanced cursor deletion in ManagedLedgerImpl**: Improved `asyncDeleteCursor()` with:
    - Proper state checking before deletion attempts
    - Separate handling for durable vs. non-durable cursors
    - Better error recovery and state management
    - Prevention of concurrent deletion attempts

12. **Enhanced waiting cursor registration**: Implemented thread-safe registration/deregistration of waiting cursors to prevent race conditions when adding or removing cursors from the waiting list, with improved debug logging.

#### Specialized Cursor Implementation Updates

13. **Non-durable cursor improvements**: Updated `NonDurableCursorImpl.asyncClose()` to properly close waiting cursors and set inactive state.

14. **Read-only cursor improvements**: Updated `ReadOnlyCursorImpl.asyncClose()` with the same proper cleanup logic.

#### Misc refactoring in ManagedCursorImpl

15. In the previous code, the `state` field was sometimes read directly and sometimes with `STATE_UPDATER.get`. Similarly, it was sometimes set directly and mostly with `STATE_UPDATER.set`. There's no need to use `STATE_UPDATER.get` and `STATE_UPDATER.set`. The volatile `state` field can be read and set directly without `STATE_UPDATER`. 
`STATE_UPDATER` is only needed for CAS operations on the `state` field.

### Key Technical Improvements

- **Eliminated infinite recursion**: The previous implementation could enter infinite recursive calls in `cancelPendingReadRequest()`, which is now resolved
- **Thread-safe operation lifecycle**: OpReadEntry instances are properly tracked through their lifecycle to prevent use-after-recycle scenarios
- **Comprehensive state management**: New state model provides better visibility and control over cursor lifecycle
- **Enhanced error handling**: Cursor closure and deletion events now properly propagate appropriate exceptions to pending operations
- **Performance optimization**: Reduced lock contention by using atomic operations instead of synchronized blocks where possible
- **Improved debugging**: Added comprehensive debug logging for cursor registration/deregistration operations
- **Robust deletion handling**: Deletion operations now properly handle failures, retries, and state recovery

### Verifying this change

- [x] Make sure that the change passes the CI checks.

This change is already covered by existing tests, specifically:
- `ReplicatorTest.testCloseReplicatorStartProducer` - verifies proper cleanup when replicator cursors are closed
- `FailoverSubscriptionTest.testWaitingCursorsCountAfterSwitchingActiveConsumers` - verifies that each cursor doesn't get added more than once into ManagedLedgerImpl's waiting cursors list
- Existing managed ledger tests that exercise concurrent read operations
- Integration tests that stress-test cursor state management under load
- Tests covering cursor deletion and state transitions

The modifications also include timeout adjustments for the test to account for the improved error handling paths.

### Impact Assessment

This change affects:
- **Threading model**: Improved thread safety in cursor state management with atomic state transitions
- **Error handling**: Better propagation of cursor closure and deactivation exceptions
- **Performance**: Reduced lock contention in high-throughput scenarios through atomic operations
- **Lifecycle management**: More robust cursor deletion and closure handling
- **State consistency**: Enhanced state model prevents invalid state transitions
- **Resource cleanup**: Better cleanup of waiting cursors and pending operations

The changes are backward compatible and do not affect public APIs or configuration defaults.

### Documentation

<!-- DO NOT REMOVE THIS SECTION. CHECK THE PROPER BOX ONLY. -->

- [ ] `doc` <!-- Your PR contains doc changes. -->
- [ ] `doc-required` <!-- Your PR changes impact docs and you will update later -->
- [x] `doc-not-needed` <!-- Your PR changes do not impact docs -->
- [ ] `doc-complete` <!-- Docs have been already added -->

lhotari commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -2958,12 +2940,18 @@ public void asyncClose(final AsyncCallbacks.CloseCallback callback, final Object
                     @Override
                     public void closeComplete(Object ctx) {
-                        STATE_UPDATER.set(ManagedCursorImpl.this, State.Closed);
+                        if (!STATE_UPDATER.compareAndSet(ManagedCursorImpl.this, State.Closing, State.Closed)) {
+                            log.warn("[{}] [{}] State was modified from closing to {} while closing", ledger.getName(),
+                                    name, state);
+                            state = State.Closed;
```
It's safe here

lhotari commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -3629,27 +3638,40 @@ void asyncDeleteCursorLedger() {
         bookkeeper.asyncDeleteLedger(cursorLedger.getId(), (rc, ctx) -> {
             ledger.mbean.endCursorLedgerDeleteOp();
             if (rc == BKException.Code.OK) {
+                state = State.Deleted;
                 log.info("[{}][{}] Deleted cursor ledger {}", ledger.getName(), name, cursorLedger.getId());
             } else {
                 log.warn("[{}][{}] Failed to delete ledger {}: {}", ledger.getName(), name, cursorLedger.getId(),
                         BKException.getMessage(rc));
                 if (!isNoSuchLedgerExistsException(rc)) {
+                    state = State.DeletingFailed;
                     ledger.getScheduledExecutor().schedule(() -> asyncDeleteCursorLedger(retry - 1),
                             DEFAULT_LEDGER_DELETE_BACKOFF_TIME_SEC, TimeUnit.SECONDS);
+                } else {
+                    state = State.Deleted;
```
sure, valid point, but STATE_UPDATE.get and get aren't providing any value over direct referencing and assignment of a volatile field.

lhotari commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -3629,27 +3638,40 @@ void asyncDeleteCursorLedger() {
                     ledger.getScheduledExecutor().schedule(() -> asyncDeleteCursorLedger(retry - 1),
                             DEFAULT_LEDGER_DELETE_BACKOFF_TIME_SEC, TimeUnit.SECONDS);
+                } else {
+                    state = State.Deleted;
```
considered

lhotari commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -3629,27 +3638,40 @@ void asyncDeleteCursorLedger() {
+        if (retry <= 0) {
+            log.warn("[{}-{}] Failed to delete ledger after retries {}", ledger.getName(), name,
+                    cursorLedger.getId());
+            state = State.DeletingFailed;
```
considered

lhotari commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -2246,7 +2228,7 @@ protected void internalAsyncMarkDelete(final Position newPosition, Map<String, L
         synchronized (pendingMarkDeleteOps) {
             // The state might have changed while we were waiting on the queue mutex
-            switch (STATE_UPDATER.get(this)) {
+            switch (state) {
```
`STATE_UPDATER.get` and `STATE_UPDATER.set` are not necessary. The volatile field can be read and set directly without `STATE_UPDATER`. The intention of the change is to use `STATE_UPDATER` only for CAS changes of the `state` field.

3pacccccc commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -1040,87 +1089,36 @@ public void asyncReadEntriesWithSkipOrWait(int maxEntries, long maxSizeBytes, Re
-                    // If there's no delay, check directly from the same thread
-                    checkForNewEntries(op, callback, ctx);
-                }
+            int opReadId = op.id;
```
I mean , We can optimize the code by avoiding unnecessary OpReadEntry.create() and OpReadEntry.recycle() operations when WAITING_READ_OP_UPDATER is not null. This improvement offers two key benefits:
- Resource Efficiency - Eliminates wasteful object creation and recycling when the operation would immediately fail
- ID Management - Since ```OpReadEntry.id``` auto-increments in this case, this may prevents ```OpReadEntry.id``` grow
rapidly and only few ```OpReadEntry.id``` are actually used for valid operations

poorbarcode commented on pulsar-broker/src/test/java/org/apache/pulsar/broker/service/ReplicatorTest.java:
```diff
@@ -1046,7 +1045,7 @@ public void testResumptionAfterBacklogRelaxed() throws Exception {
      */
-    @Test(timeOut = 15000)
+    @Test(timeOut = 30000)
     public void testCloseReplicatorStartProducer() throws Exception {
```
Start a channel to discuss the following topic

@poorbarcode left a comment

> Multiple threads attempt to cancel pending read requests concurrently
> Cursors are closed while read operations are pending
> OpReadEntry instances are recycled and reused incorrectly

Could you describe the details of how the `OpReadEntry` will be reused

---

@lhotari  left a comment

Since OpReadEntry instances are recycled, they are reused. It's not sufficient do use compareAndSet. There's a need to have some unique identifier on the instance to do the CAS operation comparison. It's possible to find code paths where this could happen and new problems could be introduced unless we have a proper solution to for CAS operations.

poorbarcode commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -1149,17 +1147,33 @@ private void checkForNewEntries(OpReadEntry op, ReadEntriesCallback callback, Ob
                     PENDING_READ_OPS_UPDATER.incrementAndGet(this);
                     ledger.asyncReadEntries(op);
                 } else {
-                    log.info("[{}] [{}] notification that new entries added was already be called, skipped the current"
```
The method `cancelPendingRead` can not cancel this task, right? which causes a orphan reading more entries

poorbarcode replied:
Okay, it will not trigger a repeated reading, just left an orphan cursor in `ml.waitingCursors`

lhotari replied:
Correct. That would get removed when the `ml.notifyCursors` method has been called.

lhotari commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -1149,17 +1147,33 @@ private void checkForNewEntries(OpReadEntry op, ReadEntriesCallback callback, Ob
-                            name, op.readPosition);
+                                name, op.readPosition);
                     }
                     PENDING_READ_OPS_UPDATER.incrementAndGet(this);
```
That's true that it could be done, however, it doesn't seem to be necessary in this solution. Calling `removeWaitingCursor` is potentially an expensive operation since `waitingCursors` is a `ConcurrentLinkedQueue` in `ManagedLedgerImpl` and when there would be a large amount of waiting cursors. That's why I'd like to avoid calling `removeWaitingCursor` here.

lhotari commented:
> > Adding `Runnable` as the parameter of `addWaitingCursorRequested` and `removeWaitingCursor` makes code a bit hard to read. When I read the code, I was wondering how many places could it be called and found only one place. IMO, it's a over-abstraction.

@BewareMyPower Yes, you are right about that it's hard to read. The reason why I didn't directly do what you are suggesting is that it's a bad idea to mutate the internal state of another instance. That breaks object encapsulation principles.

BewareMyPower commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/OpReadEntry.java:
```diff
@@ -34,7 +35,14 @@
 
 class OpReadEntry implements ReadEntriesCallback {
-
+    static final OpReadEntry WAITING_READ_OP_FOR_CLOSED_CURSOR = new OpReadEntry();
```
It can be a private static final field of `ManagedCursorImpl`.

poorbarcode commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -4012,4 +4034,55 @@ public ManagedLedgerInternalStats.CursorStats getCursorStats() {
+     * The cursor state is set to unregistered, and it can be registered again for waiting in ManagedLedgerImpl.
+     */
+    void removeWaitingCursorRequested(Runnable removeWaitingCursorRunnable) {
+        synchronized (waitingRegistrationLock) {
```
This method will be called on the publishing messages path, which is a really hot path, right?

lhotari replied:
There's no calls to `ledger.removeWaitingCursor(this)` in the hot path, so this should be fine.

BewareMyPower commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -1174,61 +1188,29 @@ private void checkForNewEntries(OpReadEntry op, ReadEntriesCallback callback, Ob
+        if (op != null) {
+            op.recycle();
         }
+        return op != null;
```
It also returns true even when `op` is `WAITING_READ_OP_FOR_CLOSED_CURSOR`. However, it means `closeWaitingCursor` has been called that the cursor has been removed from `waitingCursors`. Returning true here does not make sense.

lhotari replied:
I'll change to `op != null && op != 
OpReadEntry.WAITING_READ_OP_FOR_CLOSED_CURSOR`.

poorbarcode reviewed (approved):
After a full review, I think this PR has no bugs, and it has solved the deadlock that I explained [here](https://github.com/apache/pulsar/pull/24573#issue-3268713658)

BewareMyPower commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java:
```diff
@@ -354,6 +354,13 @@ public synchronized void removeConsumer(Consumer consumer, boolean isResetCursor
+            // read is cancelled, the dispatcher will not continue reading from the managed ledger when a new
+            // consumer is added to the dispatcher since based on the "havePendingRead" state, it will continue to
+            // expect that a read is pending and will not submit a new read.
+            topic.getManagedLedger().removeWaitingCursor(cursor);
```
It's still hard to understand the motivation here even with the comments. The comments also make me confused.

> For durable cursors

The code here does not check if the cursor is durable.

> cancel the pending read

But it does not call `cursor.cancelPendingReadRequest()`.

Is it possible to add a test?

BewareMyPower reviewed (changes requested):
I've gone through the whole PR, overall LGTM except for some coding style issues and some questions I've left. Please address these comments before merging.

P.S. The state machine is hard to read. It would be helpful to have a graph to show all possible state changes, but it's not required.

lhotari commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -791,7 +808,39 @@ private void recoveredCursor(Position position, Map<String, Long> properties,
+     * @param newState The new state to set
+     * @return The previous state of the cursor
+     */
+    State changeStateIfNotDeletingOrDeleted(State newState) {
```
True. I'll inline this method since it's only used in one location. It just adds confusion to have this method around.

lhotari commented:
> I've gone through the whole PR, overall LGTM except for some coding style issues and some questions I've left. Please address these comments before merging.
> 
> P.S. The state machine is hard to read. It would be helpful to have a graph to show all possible state changes, but it's not required.

Thank you @BewareMyPower. I've addressed your review comments. PTAL.

codelipenghui commented on pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java:
```diff
@@ -354,6 +354,16 @@ public synchronized void removeConsumer(Consumer consumer, boolean isResetCursor
+            // For non-durable cursors, there's no difference since the cursor is not expected to be used again.
+
+            // remove waiting cursor from the managed ledger, this applies to both durable and non-durable cursors.
+            topic.getManagedLedger().removeWaitingCursor(cursor);
```
Can we optimize the method of the ManagedCursor to remove from the waiting cursors if the cursor has been registered to the managed ledger? So that we can avoid call `topic.getManagedLedger().removeWaitingCursor(cursor);` here.

```
    public void setInactive() {
        if (isActive) {
            ledger.deactivateCursor(this);
            isActive = false;
        }
    }
```

lhotari replied:
> Can we optimize the method of the ManagedCursor to remove from the waiting cursors if the cursor has been registered to the managed ledger? So that we can avoid call `topic.getManagedLedger().removeWaitingCursor(cursor);` here.
> 
> ```
>     public void setInactive() {
>         if (isActive) {
>             ledger.deactivateCursor(this);
>             isActive = false;
>         }
>     }
> ```

deactivateCursor doesn't remove the cursor from waiting cursors and it shouldn't. I experimented with that and found out that it would also cause a state inconsistency with dispatcher's "havePendingRead" and make dispatchers to stop receiving messages.

lhotari replied:
> If we look back at the first issue that was fixed by #24551,
> 
> ```
> (Highlight) The method scheduleReadOnActiveConsumer resets the variable cursor.waitingReadOp, but it forgets to remove the cursor from managedledger.waitingCursors
> ```
> 
> https://github.com/apache/pulsar/pull/24551/files#diff-1d721da3aaff7689077d33bf3e55fd8d6a5c9184549857a6120d8050c5d67798L385-L386
> 
> Here, try to add another one.

Removing from the waitingCursors is a relatively costly operation due to the data structure which is a ConcurrentLinkedQueue in ManagedLedgerImpl. That is most likely why it hasn't been done in the original code. Instead of removing it, it's better to rely on ManagedLedgerImpl.notifyCursors to unqueue the waiting cursor. Therefore, the more optimal solution is to track whether there's a waiting cursor already in the queue when it has been requested.

lhotari replied:
> My concern is that we used some internal concepts from the ManagedLedger in PersistentSubscription. We should find a way to avoid that. The state of `waitingCursors` should be self-managed by ManagedLedger.

I completely agree that it should be avoided. There should have been some sort of event for handling this.
btw. The call `topic.getManagedLedger().removeWaitingCursor(cursor)` wasn't added in this PR. Before #24551, there was a similar call. The call `topic.getManagedLedger().removeWaitingCursor(cursor)` was added to `PersistentSubscription`  in #13939. Another PR in this area was #22454.

codelipenghui commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -235,6 +227,11 @@ public class ManagedCursorImpl implements ManagedCursor {
+    // This is a lock used to update the registration state of the cursor in the managed ledger.
+    private final Object waitingRegistrationLock = new Object();
+    // This is used to track if the cursor is waiting for registration in the managed ledger.
+    boolean waitingRegistered = false;
```
```suggestion
    boolean registeredToWaitingCursors = false;
```

codelipenghui commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -235,6 +227,11 @@ public class ManagedCursorImpl implements ManagedCursor {
     private volatile boolean isActive = false;
 
+    // This is a lock used to update the registration state of the cursor in the managed ledger.
+    private final Object waitingRegistrationLock = new Object();
```
```suggestion
    private final Object registerToWaitingCursorsLock = new Object();
```

BewareMyPower commented on managed-ledger/src/main/java/org/apache/bookkeeper/mledger/impl/ManagedCursorImpl.java:
```diff
@@ -3629,32 +3622,62 @@ void asyncDeleteCursorLedger() {
+        State beforeChangingState = changeStateToDeletingIfNotDeleted();
+        if (beforeChangingState == State.Deleted) {
+            log.warn("[{}-{}] Cursor ledger is already deleted. state={}", ledger.getName(), name,
+                    beforeChangingState);
```
Logging the state seems meaningless because `beforeChangingState ` is always `Deleted`.

BewareMyPower reviewed (approved):
LGTM, left a comment about logging but it's not an important issue
